modules/grpc/loki/loki-dest.cpp

```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright (c) 2023 László Várady
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "loki-dest.hpp"
+#include "loki-worker.hpp"
+
+#include "compat/cpp-start.h"
+#include "logthrdest/logthrdestdrv.h"
+#include "messages.h"
+#include "template/templates.h"
+#include "compat/cpp-end.h"
+
+#include <cstring>
+#include <string>
+#include <sstream>
+
+constexpr const auto DEFAULT_MESSAGE_TEMPLATE = "$ISODATE $HOST $MSGHDR$MSG";
+
+using syslogng::grpc::loki::DestinationDriver;
+
+struct _LokiDestDriver
+{
+  LogThreadedDestDriver super;
+  DestinationDriver *cpp;
+};
+
+DestinationDriver::DestinationDriver(LokiDestDriver *s)
+  : super(s), url("localhost:9096"), timestamp(LM_TS_PROCESSED),
+    keepalive_time(-1), keepalive_timeout(-1), keepalive_max_pings_without_data(-1)
+{
+  log_template_options_defaults(&this->template_options);
+}
+
+DestinationDriver::~DestinationDriver()
+{
+  log_template_options_destroy(&this->template_options);
+  log_template_unref(this->message);
+}
+
+void
+DestinationDriver::add_label(std::string name, LogTemplate *value)
+{
+  this->labels.push_back(Label{name, value});
+}
+
+bool
+DestinationDriver::init()
+{
+  GlobalConfig *cfg = log_pipe_get_config(&this->super->super.super.super.super);
+
+  if (!this->message)
+    {
+      this->message = log_template_new(cfg, NULL);
+      log_template_compile(this->message, DEFAULT_MESSAGE_TEMPLATE, NULL);
+    }
+
+  log_template_options_init(&this->template_options, cfg);
+
+  LogTemplate *worker_partition_key = log_template_new(cfg, NULL);
```

Comment:
There are some configurations, which do not work right now, can you please suggest, whether they are valid use-cases, and what should be the expected behavior for them?

1. No labels. Currently we get the error: `worker-partition-key() should not be literal string`
2. Static labels (`labels("foo" => "bar"))`). Currently we get the same error as in the previous point.

Maybe no labels is illogical, and we should check it and throw an error for the user, to set labels.

However setting static labels might be a valid thing to do. In this case I think we could warn the user if they have multiple `workers()` set, that we will only use 1, and we should not use the worker partition key logic of the threaded dest at all.

Reply:
I think both cases are valid and in those cases, we should just use the default round-robin worker indexing.